Added interface for GDNative Videodecoder.

zombietype opened the pull request:
Interface and callback api added for Videodecoder support.
Should be able to construct any format videodecoder using
only the given interface.

GSoC 2018 project.

zombietype committed:
Added interface for GDNative Videodecoder.

Interface and callback api added for Videodecoder support.
Should be able to construct any format videodecoder using
only the given interface.

GSoC 2018 project.

YeldhamDev commented on modules/gdnative/include/videodecoder/godot_videodecoder.h:
```diff
@@ -0,0 +1,72 @@
+/*************************************************************************/
+/*  godot_videodecoder.h                                                         */
```
A little too much spaces here.

KidRigger replied:
Fixed

LikeLakers2 replied:
Did you forget to push your fix? The PR as-is doesn't have the line fixed.

karroffel commented on modules/gdnative/videodecoder/video_stream_gdnative.cpp:
```diff
@@ -0,0 +1,336 @@
+
+	Ref<Image> img = memnew(Image(texture_size.width, texture_size.height, 0, Image::FORMAT_RGBA8, *pba));
+
+	texture->set_data(img);
```
_If_ you keep the repeating allocation, I would probably at least use `texture->set_data_partial()` here (maybe the API doesn't exist yet, but there is `VisualServer::texture_set_data_partial`)

KidRigger replied:
Should I just add the set_data_partial to ImageTexture? Just a forward to VisualServer's corresponding one.

karroffel replied:
cc @reduz

KidRigger replied:
For now committing all the fixes sans this, since webm and theora will need the fix and handling all 3 together shouldn't be much more work.

karroffel replied:
Yeah it's fine, can be changed later on, shouldn't break compatibility or anything :)

karroffel commented on modules/gdnative/videodecoder/video_stream_gdnative.cpp:
```diff
@@ -0,0 +1,336 @@
+
+Ref<VideoStreamPlayback> VideoStreamGDNative::instance_playback() {
+	Ref<VideoStreamPlaybackGDNative> pb = memnew(VideoStreamPlaybackGDNative);
+	VideoDecoderGDNative *decoder = decoder_server.get_decoder("mp4");
```
This should probably be a parameter of the function

karroffel commented on modules/gdnative/videodecoder/video_stream_gdnative.h:
```diff
@@ -0,0 +1,209 @@
+		VideoDecoderGDNative *decoder = memnew(VideoDecoderGDNative);
+		decoder->set_interface(p_interface);
+		int index = decoders.size();
+		printf("RDI\n");
```
Same here, if you have to, use `print_line` :)

karroffel replied:
boop

KidRigger replied:
This had been changed before push. My local has no printf's in the folder. I do not understand why this is showing up for you.

KidRigger commented on modules/gdnative/videodecoder/video_stream_gdnative.cpp:
```diff
@@ -0,0 +1,336 @@
+		return;
+	}
+
+	Ref<Image> img = memnew(Image(texture_size.width, texture_size.height, 0, Image::FORMAT_RGBA8, *pba));
```
Doable. I had merely used what webm seems to use.

KidRigger replied:
I'm confused as to finding a way to not repeat allocation. Because even set_data from buffer in Image class is essentially allocating Image, and copying data from the new Image.

jamie-pate replied:
Image::FORMAT_RGBA8 sucks :( it would be much more efficient to allow the plugin to choose the format

KidRigger commented on modules/gdnative/include/videodecoder/godot_videodecoder.h:
```diff
@@ -0,0 +1,72 @@
+typedef int (*GDNativeAudioMixCallback)(void *, const float *, int);
+
+typedef struct
+{
```
I added the version similar to ARVR, and a next field. (Won't that just be null?)

karroffel commented on modules/gdnative/gdnative_api.json:
```diff
@@ -11,5747 +11,9595 @@
+          [
+            "godot_color *",
+            "r_dest"
+          ],
```
Seems like you formatted a good chunk of the file, can you revert that and add your changes in the same style as the rest?

KidRigger replied:
Sigh. vscode autoformats every time I save the file. I'll change it.

KidRigger replied:
Boop comment doesn't exist in my file, and I can't even access the file. It's weird.

karroffel reviewed (commented):
Looks okay after a shallow look, just some whitespace changes that should be reverted :)

akien-mga commented on modules/gdnative/videodecoder/resource_importer_av_gdnative.cpp:
```diff
@@ -0,0 +1,96 @@
+/*************************************************************************/
+/*  resource_importer_gdnative.cpp                                       */
```
The comment should be updated to reflect this file's name (note also the trailing `*/` to close the box).

akien-mga commented on modules/gdnative/videodecoder/register_types.cpp:
```diff
@@ -0,0 +1,17 @@
+/* register_types.h */
```
For consistency, this file and its header should also use the Godot copyright header. They also lack a newline at end of file.

akien-mga commented on modules/gdnative/videodecoder/resource_importer_av_gdnative.h:
```diff
@@ -0,0 +1,58 @@
+/*************************************************************************/
+/*  resource_importer_gdnative.h                                         */
```
Same here.

akien-mga commented on modules/gdnative/videodecoder/resource_importer_av_gdnative.h:
```diff
@@ -0,0 +1,58 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef RESOURCEIMPORTERMP4FFMPEG_H
```
Might be worth renaming the guards to match the current file name/scope.

akien-mga commented on modules/gdnative/videodecoder/video_stream_gdnative.h:
```diff
@@ -0,0 +1,208 @@
+/*************************************************************************/
+/*  videostream_ffmpeg.h                                                 */
```
Same here.

KidRigger commented:
All changes made.

akien-mga commented:
@TMM @karroffel: could you review again?

akien-mga commented:
> @tmm @karroffel: could you review again?

Bump :)

karroffel commented:
GDNative side is still looking good :+1:

akien-mga commented:
@hpvb Can you check this one again? @karroffel says it's good from the GDNative end, is it also good on the feature/video decoding end?

fire commented:
Includes have been changed, so the pull request doesn't compile on master.

fire commented:
diff --git a/modules/gdnative/videodecoder/register_types.cpp b/modules/gdnative/videodecoder/register_types.cpp
index 64599365e..dfb1966a7 100644
--- a/modules/gdnative/videodecoder/register_types.cpp
+++ b/modules/gdnative/videodecoder/register_types.cpp
@@ -28,7 +28,7 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
-#include "class_db.h"
+#include "core/class_db.h"
 #include "resource_importer_av_gdnative.h"
 #include "video_stream_gdnative.h"
 
diff --git a/modules/gdnative/videodecoder/resource_importer_av_gdnative.cpp b/modules/gdnative/videodecoder/resource_importer_av_gdnative.cpp
index ff9f6118f..7fe8cc89d 100644
--- a/modules/gdnative/videodecoder/resource_importer_av_gdnative.cpp
+++ b/modules/gdnative/videodecoder/resource_importer_av_gdnative.cpp
@@ -30,8 +30,8 @@
 
 #include "resource_importer_av_gdnative.h"
 
-#include "io/resource_saver.h"
-#include "os/file_access.h"
+#include "core/io/resource_saver.h"
+#include "core/os/file_access.h"
 #include "scene/resources/texture.h"
 
 String ResourceImporterAVGDNative::get_importer_name() const {
diff --git a/modules/gdnative/videodecoder/video_stream_gdnative.cpp b/modules/gdnative/videodecoder/video_stream_gdnative.cpp
index a959a4c8a..2e2c400d5 100644
--- a/modules/gdnative/videodecoder/video_stream_gdnative.cpp
+++ b/modules/gdnative/videodecoder/video_stream_gdnative.cpp
@@ -29,7 +29,7 @@
 /*************************************************************************/
 
 #include "video_stream_gdnative.h"
-#include <project_settings.h>
+#include <core/project_settings.h>
 #include <servers/audio_server.h>
 
 VideoDecoderServer *VideoDecoderServer::instance = NULL;
diff --git a/modules/gdnative/videodecoder/video_stream_gdnative.h b/modules/gdnative/videodecoder/video_stream_gdnative.h
index d203a5d04..ee9410e3c 100644
--- a/modules/gdnative/videodecoder/video_stream_gdnative.h
+++ b/modules/gdnative/videodecoder/video_stream_gdnative.h
@@ -32,7 +32,7 @@
 #define VIDEO_STREAM_GDNATIVE_H
 
 #include <modules/gdnative/gdnative.h>
-#include <os/file_access.h>
+#include <core/os/file_access.h>
 #include <scene/resources/texture.h>
 #include <scene/resources/video_stream.h>

akien-mga committed:
Fix includes and initialization for GDNative Videodecoder

Fixes warnings and a crash when running the destructor with an
uninitialized pcm pointer.

akien-mga commented:
I rebased and pushed a new commit to this PR to fix includes, some style issues and a compile warning.

There's one thing left to do before merging though, as `VideoStreamGDNative` was based off `VideoStream{Theora,Webm}` as some point in the past and lacks bug fixes done recently: #19937 and #21084.

akien-mga committed:
Adapt GDNative VideoStream to use ResourceFormatLoader

Same logic as used for Webm and Theora support in #19937 and #21084,
fixing issues when exporting projects.

akien-mga commented:
I got this crash when creating a new `VideoStreamGDNative` as stream in the Inspector for a `VideoPlayer` node:
```
[1] /lib64/libc.so.6(+0x3a120) [0x7fd5d0330120] (??:0)
[2] Memory::free_static(void*, bool) (/home/akien/Projects/godot/godot.git/core/os/memory.cpp:178)
[3] VideoStreamPlaybackGDNative::cleanup() (/home/akien/Projects/godot/godot.git/modules/gdnative/videodecoder/video_stream_gdnative.cpp:203)
[4] VideoStreamPlaybackGDNative::~VideoStreamPlaybackGDNative() (/home/akien/Projects/godot/godot.git/modules/gdnative/videodecoder/video_stream_gdnative.cpp:195)
[5] void memdelete<VideoStreamPlaybackGDNative>(VideoStreamPlaybackGDNative*) (/home/akien/Projects/godot/godot.git/./core/os/memory.h:120)
[6] Ref<VideoStreamPlaybackGDNative>::unref() (/home/akien/Projects/godot/godot.git/./core/reference.h:271)
[7] Ref<VideoStreamPlaybackGDNative>::~Ref() (/home/akien/Projects/godot/godot.git/./core/reference.h:286)
[8] VideoStreamGDNative::instance_playback() (/home/akien/Projects/godot/godot.git/modules/gdnative/videodecoder/video_stream_gdnative.cpp:305)
[9] VideoPlayer::set_stream(Ref<VideoStream> const&) (/home/akien/Projects/godot/godot.git/scene/gui/video_player.cpp:209)
[10] MethodBind1<Ref<VideoStream> const&>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/./core/method_bind.gen.inc:729 (discriminator 12))
[11] ClassDB::set_property(Object*, StringName const&, Variant const&, bool*) (/home/akien/Projects/godot/godot.git/core/class_db.cpp:1020)
[12] Object::set(StringName const&, Variant const&, bool*) (/home/akien/Projects/godot/godot.git/core/object.cpp:422)
[13] UndoRedo::_process_operation_list(List<UndoRedo::Operation, DefaultAllocator>::Element*) (/home/akien/Projects/godot/godot.git/core/undo_redo.cpp:285 (discriminator 2))
[14] UndoRedo::redo() (/home/akien/Projects/godot/godot.git/core/undo_redo.cpp:311)
[15] UndoRedo::commit_action() (/home/akien/Projects/godot/godot.git/core/undo_redo.cpp:249)
[16] EditorInspector::_edit_set(String const&, Variant const&, bool, String const&) (/home/akien/Projects/godot/godot.git/editor/editor_inspector.cpp:1965)
[17] EditorInspector::_property_changed(String const&, Variant const&, bool) (/home/akien/Projects/godot/godot.git/editor/editor_inspector.cpp:1979 (discriminator 2))
[18] MethodBind3<String const&, Variant const&, bool>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/./core/method_bind.gen.inc:2151 (discriminator 26))
[19] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/core/object.cpp:945 (discriminator 1))
[20] Object::emit_signal(StringName const&, Variant const**, int) (/home/akien/Projects/godot/godot.git/core/object.cpp:1231 (discriminator 1))
[21] Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/core/object.cpp:1287)
[22] EditorPropertyResource::_menu_option(int) (/home/akien/Projects/godot/godot.git/editor/editor_properties.cpp:2170 (discriminator 8))
[23] MethodBind1<int>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/./core/method_bind.gen.inc:729 (discriminator 12))
[24] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/core/object.cpp:945 (discriminator 1))
[25] Object::emit_signal(StringName const&, Variant const**, int) (/home/akien/Projects/godot/godot.git/core/object.cpp:1231 (discriminator 1))
[26] Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/core/object.cpp:1287)
[27] PopupMenu::activate_item(int) (/home/akien/Projects/godot/godot.git/scene/gui/popup_menu.cpp:1102 (discriminator 4))
[28] PopupMenu::_gui_input(Ref<InputEvent> const&) (/home/akien/Projects/godot/godot.git/scene/gui/popup_menu.cpp:330)
[29] MethodBind1<Ref<InputEvent> const&>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/./core/method_bind.gen.inc:729 (discriminator 12))
[30] Object::call_multilevel(StringName const&, Variant const**, int) (/home/akien/Projects/godot/godot.git/core/object.cpp:779 (discriminator 1))
[31] Object::call_multilevel(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/core/object.cpp:886)
[32] Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) (/home/akien/Projects/godot/godot.git/scene/main/viewport.cpp:1514 (discriminator 2))
[33] Viewport::_gui_input_event(Ref<InputEvent>) (/home/akien/Projects/godot/godot.git/scene/main/viewport.cpp:1854 (discriminator 3))
[34] Viewport::input(Ref<InputEvent> const&) (/home/akien/Projects/godot/godot.git/scene/main/viewport.cpp:2603 (discriminator 2))
[35] Viewport::_vp_input(Ref<InputEvent> const&) (/home/akien/Projects/godot/godot.git/scene/main/viewport.cpp:1291)
[36] MethodBind1<Ref<InputEvent> const&>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/./core/method_bind.gen.inc:729 (discriminator 12))
[37] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/core/object.cpp:945 (discriminator 1))
[38] Object::call(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/core/object.cpp:870)
[39] SceneTree::call_group_flags(unsigned int, StringName const&, StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/scene/main/scene_tree.cpp:262)
[40] SceneTree::input_event(Ref<InputEvent> const&) (/home/akien/Projects/godot/godot.git/scene/main/scene_tree.cpp:418 (discriminator 6))
[41] InputDefault::_parse_input_event_impl(Ref<InputEvent> const&, bool) (/home/akien/Projects/godot/godot.git/main/input_default.cpp:414)
[42] InputDefault::parse_input_event(Ref<InputEvent> const&) (/home/akien/Projects/godot/godot.git/main/input_default.cpp:260)
[43] OS_X11::process_xevents() (/home/akien/Projects/godot/godot.git/platform/x11/os_x11.cpp:1997 (discriminator 3))
[44] OS_X11::run() (/home/akien/Projects/godot/godot.git/platform/x11/os_x11.cpp:2819)
[45] /home/akien/Projects/godot/godot.git/bin/godot.x11.tools.64(main+0xbb) [0x119b30d] (/home/akien/Projects/godot/godot.git/platform/x11/godot_x11.cpp:56)
[46] /lib64/libc.so.6(__libc_start_main+0xeb) [0x7fd5d031a02b] (??:0)
[47] /home/akien/Projects/godot/godot.git/bin/godot.x11.tools.64(_start+0x2a) [0x119b1aa] (/home/iurt/rpmbuild/BUILD/glibc-2.28/csu/../sysdeps/x86_64/start.S:122)
```

akien-mga commented:
012dac9 fixes the crash and the error for me.

akien-mga commented on modules/gdnative/videodecoder/video_stream_gdnative.cpp:
```diff
@@ -198,7 +209,8 @@ VideoStreamPlaybackGDNative::~VideoStreamPlaybackGDNative() {
 	if (data_struct)
 		interface->destructor(data_struct);
-	memfree(pcm);
+	if (pcm)
```
Without this check, we got:
```
ERROR: free_static: Condition ' p_ptr == __null ' is true.
   At: core/os/memory.cpp:161.
```

akien-mga commented:
> > There's one thing left to do before merging though, as `VideoStreamGDNative` was based off `VideoStream{Theora,Webm}` as some point in the past and lacks bug fixes done recently: #19937 and #21084.
> 
> I've done the relevant work in [akien-mga@dbb5cb6](https://github.com/akien-mga/godot/commit/dbb5cb6dc035c3397e1ed865542971b9d8e3f3c2)
> I'm not adding it to this PR yet as I think the crash reported above should be investigated first.

Now that the crash is fixed, I've added this to this PR too: 72feccf20.

Still untested with the actual GDNative plugin as I'm not sure how to use it :)

akien-mga commented:
> Now that the crash is fixed, I've added this to this PR too: 72feccf.
> 
> Still untested with the actual GDNative plugin as I'm not sure how to use it :)

Well, got a crash(*) with this commit and the https://github.com/KidRigger/godot-videodecoder test project, so I'll remove it for now until we get things working with the original state. (PR back to 012dac9.)

I can't get the test project to load though with 012dac9, I get this error:
```
ERROR: _load: No loader found for resource: res://samples/big_buck_bunny.mp4
   At: core/io/resource_loader.cpp:192.
ERROR: _load: No loader found for resource: res://samples/big_buck_bunny.mp4
   At: core/io/resource_loader.cpp:192.
WARNING: _parse_ext_resource: Couldn't load external resource: res://samples/big_buck_bunny.mp4
   At: scene/resources/scene_format_text.cpp:174.
ERROR: remove_scene: Index p_idx=1 out of size (edited_scene.size()=1)
   At: editor/editor_data.cpp:564.
```
I did build the Linux .so with `scons platform=linux`.

-----

(*) Backtrace:
<details>
<pre>
[1] /lib64/libc.so.6() [0x311aa3a120] (??:0)
[2] void memdelete<FileAccess>(FileAccess*) (/home/akien/Projects/godot/godot.git/./core/os/memory.h:119)
[3] ResourceFormatLoaderVideoStreamGDNative::load(String const&, String const&, Error*) (/home/akien/Projects/godot/godot.git/modules/gdnative/videodecoder/video_stream_gdnative.cpp:359)
[4] ResourceLoader::_load(String const&, String const&, String const&, bool, Error*) (/home/akien/Projects/godot/godot.git/core/io/resource_loader.cpp:179 (discriminator 6))
[5] ResourceLoader::load(String const&, String const&, bool, Error*) (/home/akien/Projects/godot/godot.git/core/io/resource_loader.cpp:243)
[6] ResourceInteractiveLoaderText::poll() (/home/akien/Projects/godot/godot.git/scene/resources/scene_format_text.cpp:434)
[7] ResourceFormatLoader::load(String const&, String const&, Error*) (/home/akien/Projects/godot/godot.git/core/io/resource_loader.cpp:144)
[8] ResourceLoader::_load(String const&, String const&, String const&, bool, Error*) (/home/akien/Projects/godot/godot.git/core/io/resource_loader.cpp:179 (discriminator 6))
[9] ResourceLoader::load(String const&, String const&, bool, Error*) (/home/akien/Projects/godot/godot.git/core/io/resource_loader.cpp:243)
[10] EditorNode::load_scene(String const&, bool, bool, bool, bool) (/home/akien/Projects/godot/godot.git/editor/editor_node.cpp:2864 (discriminator 2))
[11] EditorNode::_sources_changed(bool) (/home/akien/Projects/godot/godot.git/editor/editor_node.cpp:535)
[12] MethodBind1<bool>::call(Object*, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/./core/method_bind.gen.inc:729 (discriminator 12))
[13] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/akien/Projects/godot/godot.git/core/object.cpp:945 (discriminator 1))
[14] Object::emit_signal(StringName const&, Variant const**, int) (/home/akien/Projects/godot/godot.git/core/object.cpp:1231 (discriminator 1))
[15] Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/akien/Projects/godot/godot.git/core/object.cpp:1287)
[16] EditorFileSystem::_notification(int) (/home/akien/Projects/godot/godot.git/editor/editor_file_system.cpp:1102 (discriminator 4))
[17] EditorFileSystem::_notificationv(int, bool) (/home/akien/Projects/godot/godot.git/editor/editor_file_system.h:107 (discriminator 14))
[18] Object::notification(int, bool) (/home/akien/Projects/godot/godot.git/core/object.cpp:957)
[19] SceneTree::_notify_group_pause(StringName const&, int) (/home/akien/Projects/godot/godot.git/scene/main/scene_tree.cpp:967)
[20] SceneTree::idle(float) (/home/akien/Projects/godot/godot.git/scene/main/scene_tree.cpp:510 (discriminator 3))
[21] Main::iteration() (/home/akien/Projects/godot/godot.git/main/main.cpp:1863)
[22] OS_X11::run() (/home/akien/Projects/godot/godot.git/platform/x11/os_x11.cpp:2821)
[23] godot-git(main+0xbb) [0x119a5ad] (/home/akien/Projects/godot/godot.git/platform/x11/godot_x11.cpp:56)
[24] /lib64/libc.so.6(__libc_start_main+0xeb) [0x311aa2402b] (??:0)
[25] godot-git(_start+0x2a) [0x119a44a] (/home/iurt/rpmbuild/BUILD/glibc-2.28/csu/../sysdeps/x86_64/start.S:122)
</pre>
</details>

KidRigger commented:
Test project needs the samples downloaded. Are you missing the Samples module in the test folder?

If not then this issue is new and unknown. I'll catch up to all the changes and get back to this issue.

akien-mga commented:
>> I can't get the test project to load though with 012dac9
>
> Test project needs the samples downloaded. Are you missing the Samples module in the test folder?

I did forget to pull the submodule, I'll try again. Might be worth mentioning it in the repo's README.md, Git is not very explicit about the existence of submodules when you clone a repo.

akien-mga commented:
Tested again with 012dac9 and the properly checked out samples, I now get a crash when opening the editor:
```
ctor()
set_audio_track()
open_file()
QuickTime / MOV

Thread 1 "godot-git" received signal SIGSEGV, Segmentation fault.
godot_videodecoder_open_file (p_data=0xc6d5d30, file=0xfa239a0) at src/gdnative_videodecoder.c:414
414                     if (data->format_ctx->streams[i]->codecpar->codec_type == AVMEDIA_TYPE_VIDEO) {

(gdb) bt
#0  godot_videodecoder_open_file (p_data=0xc6d5d30, file=0xfa239a0) at src/gdnative_videodecoder.c:414
#1  0x000000000123cfb4 in VideoStreamPlaybackGDNative::open_file (this=0x63d9920, p_file=...) at modules/gdnative/videodecoder/video_stream_gdnative.cpp:118
#2  0x000000000123de52 in VideoStreamGDNative::instance_playback (this=0xcf8f670) at modules/gdnative/videodecoder/video_stream_gdnative.cpp:322
#3  0x00000000026a9b9c in VideoPlayer::set_stream (this=0xfa1cda0, p_stream=...) at scene/gui/video_player.cpp:209
#4  0x00000000026af957 in MethodBind1<Ref<VideoStream> const&>::call (this=0x593a330, p_object=0xfa1cda0, p_args=0x7fffffffbcb8, p_arg_count=1, r_error=...)
    at ./core/method_bind.gen.inc:729
#5  0x00000000030817c5 in ClassDB::set_property (p_object=0xfa1cda0, p_property=..., p_value=..., r_valid=0x7fffffffbecf) at core/class_db.cpp:1020
#6  0x00000000030f99d8 in Object::set (this=0xfa1cda0, p_name=..., p_value=..., r_valid=0x7fffffffbecf) at core/object.cpp:422
#7  0x0000000002ba4577 in SceneState::instance (this=0x61c3060, p_edit_state=SceneState::GEN_EDIT_STATE_MAIN) at scene/resources/packed_scene.cpp:252
#8  0x0000000002baeaf6 in PackedScene::instance (this=0xfa177d0, p_edit_state=PackedScene::GEN_EDIT_STATE_MAIN) at scene/resources/packed_scene.cpp:1697
#9  0x0000000001c5a27e in EditorNode::load_scene (this=0x65c89b0, p_scene=..., p_ignore_broken_deps=false, p_set_inherited=false, p_clear_errors=true, p_force_open_imported=false)
    at editor/editor_node.cpp:2919
#10 0x0000000001c43fb6 in EditorNode::_sources_changed (this=0x65c89b0, p_exist=false) at editor/editor_node.cpp:534
#11 0x0000000001336524 in MethodBind1<bool>::call (this=0xb29cde0, p_object=0x65c89b0, p_args=0x7fffffffc8c0, p_arg_count=1, r_error=...) at ./core/method_bind.gen.inc:729
#12 0x00000000030fc60d in Object::call (this=0x65c89b0, p_method=..., p_args=0x7fffffffc8c0, p_argcount=1, r_error=...) at core/object.cpp:945
#13 0x00000000030fde96 in Object::emit_signal (this=0x6d17380, p_name=..., p_args=0x7fffffffc8c0, p_argcount=1) at core/object.cpp:1231
#14 0x00000000030fe505 in Object::emit_signal (this=0x6d17380, p_name=..., p_arg1=..., p_arg2=..., p_arg3=..., p_arg4=..., p_arg5=...) at core/object.cpp:1287
#15 0x0000000001beb135 in EditorFileSystem::_notification (this=0x6d17380, p_what=17) at editor/editor_file_system.cpp:1102
#16 0x0000000001bf3488 in EditorFileSystem::_notificationv (this=0x6d17380, p_notification=17, p_reversed=false) at editor/editor_file_system.h:107
#17 0x00000000030fc78a in Object::notification (this=0x6d17380, p_notification=17, p_reversed=false) at core/object.cpp:955
#18 0x000000000247b96e in SceneTree::_notify_group_pause (this=0x5d92be0, p_group=..., p_notification=17) at scene/main/scene_tree.cpp:967
#19 0x00000000024796ec in SceneTree::idle (this=0x5d92be0, p_time=0.150000095) at scene/main/scene_tree.cpp:510
#20 0x00000000011d38ff in Main::iteration () at main/main.cpp:1862
#21 0x00000000011a8a29 in OS_X11::run (this=0x7fffffffd230) at platform/x11/os_x11.cpp:2821
#22 0x000000000119b38d in main (argc=2, argv=0x7fffffffd968) at platform/x11/godot_x11.cpp:55
```

Note that I'm compiling against system-installed ffmpeg 4.1.

akien-mga commented:
I re-added 72feccf2, after discussing with @KidRigger it seems the issues I had were likely due to how I built the GDNative plugin with system ffmpeg, but the engine-side interface still seems to be fine.

Let's merge it as is and refine from here.

akien-mga commented:
Thanks a ton for your work :) Looking forward to the next steps!
